Recipe app

amexabee opened the pull request:
- Login page and registration page:
  - Use Devise
- Food list:
  - Displays a list of food added by the logged-in user 
  - Displays also quantity of a given food
  - Leads to a form that allows users to add new food
- Recipes list:
  - Displays a list of recipes created by the logged-in user 
  - Leads to recipe details
  - If the user is the owner of the recipe, allow the user to delete it
- Public recipe list:
  - Displays a list of all public recipes ordered by the newest 
  - Lead to recipe details.
- Recipe details:
  - Display a toggle button that allows for a recipe to be made public or private.
  - If the recipe is public or the user is the owner of the recipe, display the recipe details 
  - If the user is the owner of the recipe, lead to the form that allows the user to add new food.
- General shopping list view 
  - List food that is missing for all recipes of the logged-in user 
  - Count the total food items and total price of the missing food
- Make sure there are no N+1 queries happening
- Create a navigation menu that allows users to open all of the pages we created
- Write unit and integration tests

amexabee committed:
Create README.md

amexabee committed:
Update README.md

amexabee committed:
Create LICENSE

lmunozm1702 committed:
Install rails

lmunozm1702 committed:
Add gems

lmunozm1702 committed:
Add microverse linters configuration.

lmunozm1702 committed:
Fix rubocop offenses

lmunozm1702 committed:
Add bootstrap 5

lmunozm1702 committed:
Fix linter errors.

lmunozm1702 committed:
Merge pull request #18 from lmunozm1702/setup_project

Setup project

lmunozm1702 committed:
Install and configure devise

- Create user table, and run the migration.
- Setup devise gem.

lmunozm1702 committed:
Remove turbo-rails, active_storage and action_mailbox routes

lmunozm1702 committed:
Set / to devise login page.

lmunozm1702 committed:
Fix rubocop offences

lmunozm1702 committed:
Update users.rb

lmunozm1702 committed:
Merge pull request #19 from lmunozm1702/login_with_devise

Login with devise

lmunozm1702 committed:
Create Food model and components.

lmunozm1702 committed:
Add food specs

amexabee committed:
Create Recipe model

amexabee committed:
Create Recipes controller

amexabee committed:
Display a list of recipes

amexabee committed:
Change preparation and cooking time data type

lmunozm1702 committed:
Add Food CRUD

- Without update as required.

lmunozm1702 committed:
Solve Rubocop errors

lmunozm1702 committed:
Merge pull request #20 from lmunozm1702/food_crud

Food CRUD

amexabee committed:
Add simple view for recipe detail

amexabee committed:
Add basic deleting functionality

amexabee committed:
Fix linters

amexabee committed:
Merge branch 'dev' into recipe-crud

amexabee committed:
Merge pull request #21 from lmunozm1702/recipe-crud

Recipe crud

lmunozm1702 committed:
Add list_all_foods into food model.

lmunozm1702 committed:
Merge pull request #25 from lmunozm1702/list_all_foods

Add list_all_foods into food model.

amexabee committed:
Add new recipe creating logic

lmunozm1702 committed:
Add navbar

lmunozm1702 committed:
Merge pull request #26 from lmunozm1702/navbar

Add Navbar

lmunozm1702 committed:
Update Foods

- change input type from text to select for Measurement unit.

lmunozm1702 committed:
Style devise views

lmunozm1702 committed:
Merge pull request #27 from lmunozm1702/devise_styles

Devise styles

amexabee committed:
Add more to view of recipe detail

amexabee committed:
Add Generate shop list and add ingredient buttons

amexabee committed:
Fix linters

amexabee committed:
Merge pull request #28 from lmunozm1702/recipe-detail

Recipe detail

amexabee committed:
Fix breaking changes

amexabee committed:
Add public recipes, update recipes list

amexabee committed:
Fix linters

amexabee committed:
Modify public recipes view to match wireframe

amexabee committed:
Merge pull request #29 from lmunozm1702/public-recipes

Public recipes

amexabee committed:
Add Recipe Food model

amexabee committed:
Complete logic for adding ingredient

amexabee committed:
Fix linters

amexabee committed:
Apply reviewrs comment

amexabee committed:
Add ingredient removing functionality

amexabee committed:
Fix linters

amexabee committed:
Merge pull request #30 from lmunozm1702/recipe-food

Recipe food

amexabee committed:
Add rails_helper and spec_helper

amexabee committed:
Add tests for recipe and recipe_food model

amexabee committed:
Add tests for recipes, public_recipes and recipe_foods controllers

amexabee committed:
Add Integration tests

lmunozm1702 committed:
Create shopping list view

lmunozm1702 committed:
Add database.yml to gitignore

amexabee committed:
Fix linters

lmunozm1702 committed:
Fix rubocop issues

lmunozm1702 committed:
Activate "Generate Shopping List" button from Recipe Details.

lmunozm1702 committed:
Merge pull request #31 from lmunozm1702/shopping_list

Shopping list

amexabee committed:
Add more integration tests

amexabee committed:
Add spec for food controller

lmunozm1702 committed:
Add general shopping list

Update navbar to activate "General Shopping List" option
create general shopping list view.

amexabee committed:
Add spec for user and food models

amexabee committed:
Add spec for user model

lmunozm1702 committed:
Merge pull request #32 from lmunozm1702/general_shopping_list

Add general shopping list

amexabee committed:
Fix linters

lmunozm1702 committed:
Add new recipe to Navbar

amexabee committed:
Remove a file from PR

lmunozm1702 committed:
Style new recipe.

lmunozm1702 committed:
Style add ingredient view.

amexabee committed:
Merge pull request #33 from lmunozm1702/feature-spec

Feature spec

lmunozm1702 committed:
Style recipes index

lmunozm1702 committed:
Style Public Recipes view

lmunozm1702 committed:
Merge pull request #34 from lmunozm1702/styling

Styling Recipes Views

amexabee committed:
Solve edge cases

amexabee committed:
Solve edge cases

rloterh commented on app/views/foods/_form.html.erb:
```diff
@@ -0,0 +1,34 @@
+<%= form_with(model: food) do |form| %>
```
- [ ] There is mismatch in the _price_ and _quantity_ values for _add food item_ and the values displayed on the table. Kindly ensure that values entered match accordingly with the view:
**Improve your design:**

**Form:**        |  **Table:**        |  **Expected:**     
:-------------------------:|:-------------------------:|:-------------------------:
<img width="800" alt="Home Page" src="https://user-images.githubusercontent.com/12745474/226095048-4597b2d5-db80-4c7d-95e7-ca2d7bc0d0ff.png"> | <img width="800" alt="Home Page" src="https://user-images.githubusercontent.com/12745474/226095518-cd59bf8e-5754-4bad-a63c-2b03b4faeccb.png"> | <img width="800" alt="Home Page" src="https://user-images.githubusercontent.com/12745474/226095562-19e6bc38-3743-43a4-ac94-900cf3c1aa95.png">

lmunozm1702 replied:
Hi @rloterh, please help with this issue, the entered number is 3 into the Quantity field in the New Food form, and the signed value into the table is 3 too. what is the value expected?

lmunozm1702 replied:
Solved, I change the value to the correct column "price"

![image](https://user-images.githubusercontent.com/53025062/226100365-e091998d-d61c-4989-88d8-b94ed031641c.png)

rloterh commented on app/views/recipe_foods/general_shopping.html.erb:
```diff
@@ -0,0 +1,29 @@
+<%= render partial: 'layouts/section_title', locals: { section_title: 'General Shopping List' } %>
```
- [x] [OPTIONAL] General shopping list view is throwing an error. I suggest you remove it as this feature is required by a group of 3 contributors.

**Page error:**
![image](https://user-images.githubusercontent.com/12745474/226096538-5b4f7784-8cf1-4ed8-a42d-795b7f1b399e.png)

lmunozm1702 replied:
Option removed as required.

rloterh commented on app/models/user.rb:
```diff
@@ -0,0 +1,9 @@
+  # :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable,
+         :confirmable
```
- [x] [OPTIONAL] It was difficult to sign in after sign up as email confirmation was not working as intended. I had had to temporary disable this option in order to sign in successfully. Kindly improve upon this aspect.

lmunozm1702 replied:
Devise send the same email needed to activate the account or in the new password flow, to the server log, there we can click the link to finish the signup process.

![image](https://user-images.githubusercontent.com/53025062/226099149-9f3e5015-88be-4fc5-ace4-5fba022315cb.png)

rloterh commented on spec/models/food_spec.rb:
```diff
@@ -0,0 +1,28 @@
+      @food2 = Food.create(name: 'Doro', user_id: @user.id, unit: 'grams', quantity: 1)
+      foods = Food.list_all_foods
+
+      expect(foods).to eq(foods.reverse)
```
- [x] Few of the test specs are failing. Kindly ensure all test cases are passing.

**Failing test cases:**
![image](https://user-images.githubusercontent.com/12745474/226097327-e1d549a9-3af2-4ce2-b64d-718d5a8843af.png)

lmunozm1702 replied:
Solved.

rloterh reviewed (changes requested):
Hi Luis & Amanuel, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 🚦 🚦  🏃   

## Highlights 👍  
- App schema closely follows given ERD!
- Unit test framework properly setup!
- Professional ReadMe!
- Reviewed projects in turns!

## Required Changes ♻️
_Kindly check the inline comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

_Kindly consider my inline comments marked **[OPTIONAL]** under this review_

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

lmunozm1702 committed:
Solve Code Review Issue

Remove Menu Item: "General Shopping List"

Optional Issue: General shopping list view is throwing an error. I suggest you remove it as this feature is required by a group of 3 contributors.

lmunozm1702 committed:
Solve code review issue - Rspec tests

Issue:  Few of the test specs are failing. Kindly ensure all test cases are passing.

lmunozm1702 committed:
Solve code review issue

There is mismatch in the price and quantity values for add food item and the values displayed on the table. Kindly ensure that values entered match accordingly with the view:

amexabee committed:
Solve edge cases

amexabee committed:
Clear conflict

amexabee committed:
Revert changes made to database.yml

AnselemOdims reviewed (approved):
### Hi  @team🙋‍♀️, 
Good job in making the required changes 👏 

# STATUS APPROVED ✔
Your project is complete!  There is nothing else to say other than... it's time to merge it 🔗 🚢 
Congratulations! 🎉 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers 🥂  and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Use the tag @AnselemOdims for a quicker reply


**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
